ubuntu/updaterset.go

```diff
@@ -2,125 +2,307 @@ package ubuntu
+	ch := make(chan *distroSeries)
+	us := make(chan *updater)
+	eg.Go(func() error {
+		// Send active distribution series down the channel.
+		defer close(ch)
+		for i := range series.Entries {
+			e := &series.Entries[i]
+			mkDist(e.Version, e.Name)
+			if !e.Active {
+				zlog.Debug(ctx).Str("release", e.Name).Msg("release not active")
+				continue
+			}
+			select {
+			case ch <- e:
+			case <-ctx.Done():
+				return ctx.Err()
+			}
+		}
+		return nil
+	})
+	eg.Go(func() error {
+		// Double-check the distribution.
+		defer close(us)
+		for e := range ch {
+			url := fmt.Sprintf("https://security-metadata.canonical.com/oval/com.ubuntu.%s.cve.oval.xml", e.Name)
+			req, err := http.NewRequestWithContext(ctx, http.MethodHead, url, nil)
+			if err != nil {
+				return fmt.Errorf("ubuntu: unable to construct request: %w", err)
+			}
+			req.Header.Set(`accept`, `application/x-bzip2,application/xml;q=0.9`)
+			res, err := f.c.Do(req)
+			if err != nil {
+				return fmt.Errorf("ubuntu: error requesting inspecting OVAL feed: %w", err)
+			}
+			defer res.Body.Close()
+			switch res.StatusCode {
+			case http.StatusOK:
+			case http.StatusNotFound:
+				zlog.Debug(ctx).
+					Str("name", e.Name).
+					Str("version", e.Version).
+					Msg("OVAL database missing, skipping")
+				continue
+			default:
+				return fmt.Errorf("ubuntu: unexpected status requesting %q: %q", url, res.Status)
+			}
+			next, err := res.Request.URL.Parse(res.Header.Get(`content-location`))
+			if err != nil {
+				// ???
+			}
+			us <- &updater{
+				url:      next.String(),
+				useBzip2: strings.EqualFold(`application/x-bzip2`, res.Header.Get(`content-type`)),
+				name:     e.Name,
+				id:       e.Version,
+			}
+		}
+		return nil
+	})
+	eg.Go(func() error {
+		//Construct the set
+		for u := range us {
+			if err := set.Add(u); err != nil {
+				return err
+			}
+		}
+		return nil
+	})
+	if err := eg.Wait(); err != nil {
 		return set, err
 	}
-	for _, u := range us {
-		if u == nil {
-			continue
-		}
-		if err := set.Add(u); err != nil {
-			return set, err
+
+	if len(f.force) != 0 {
+		zlog.Info(ctx).Msg("configuring manually specified updaters")
+		ns := make([]string, 0, len(f.force))
+		for _, p := range f.force {
+			u := &updater{
+				url:      fmt.Sprintf("https://security-metadata.canonical.com/oval/com.ubuntu.%s.cve.oval.xml.bz2", p[0]),
+				useBzip2: true,
+				name:     p[0],
+				id:       p[1],
+			}
+			if err := set.Add(u); err != nil {
+				// Already exists, skip.
+				zlog.Debug(ctx).Err(err).Msg("skipping updater")
+				continue
+			}
+			ns = append(ns, u.Name())
 		}
+		zlog.Info(ctx).Strs("updaters", ns).Msg("added specified updaters")
 	}
+
 	return set, nil
 }
+
+type seriesResponse struct {
+	TotalSize        int            `json:"total_size"`
+	Start            int            `json:"start"`
+	Entries          []distroSeries `json:"entries"`
+	ResourceTypeLink string         `json:"resource_type_link"`
+}
+
+type distroSeries struct {
```

Comment:
This is kind of a nit, but is there a reason to define all these fields and not just `Active`, `Name` and `Version` (there might be others needed that I've missed)

Reply:
No, these are just auto-generated from a live response. I'll bust them down.